### PR TITLE
Implement Encodable for vec of references of an Encodable object

### PR DESCRIPTION
**Pull request by jrawsthorne**
Allows for serialization of Vec<&TxOut> as an example. Happy to duplicate the code instead of extracting out to a function if preferred as there isn't much to it.

### CODE REVIEW

File: src/consensus/encode.rs [hunk 1]
```diff
@@ -545,21 +545,35 @@ impl Encodable for [u16; 8] {
...
-                mut s: S,
+                s: S,
             ) -> Result<usize, io::Error> {
-                let mut len = 0;
-                len += VarInt(self.len() as u64).consensus_encode(&mut s)?;
-                for c in self.iter() {
-                    len += c.consensus_encode(&mut s)?;
-                }
-                Ok(len)
+                encode_vec(s, &self)
```
Comment (dr-orlovsky) [inline src/consensus/encode.rs:567]: Not sure why we are taking self by reference here in explicit form; omitting `&` should also work (as it assumes reference from the `&self` in method argument)

File: src/consensus/encode.rs [hunk 2]
```diff
@@ -545,21 +545,35 @@ impl Encodable for [u16; 8] {
...
+            }
+        }
+        impl<'a> Encodable for Vec<&'a $type> {
+            #[inline]
+            fn consensus_encode<S: io::Write>(
+                &self,
+                s: S,
+            ) -> Result<usize, io::Error> {
+                encode_vec(s, &self)
```
Comment (dr-orlovsky) [inline src/consensus/encode.rs:576]: Ibid

**Review by dr-orlovsky — DISMISSED** [review]
I would prefer to use iterator instead of `vec`, but AFAIR there is no other non-vec types which are used in consensus encoding, so non-iterating impl is also fine

File: src/consensus/encode.rs [hunk 3]
```diff
@@ -545,21 +545,35 @@ impl Encodable for [u16; 8] {
     }
 }
 
+#[inline]
+fn encode_vec<S: io::Write, T: Encodable>(mut s: S, vec: &Vec<T>) -> Result<usize, io::Error> {
```
Comment (sgeisler) [inline src/consensus/encode.rs:549]: Dumb question, but why even a `&Vec<T>` instead of `&[T]`?

**Review by sgeisler — COMMENTED** [review]
I think once you begin to borrow `Vec`s you should nearly always go with slices instead. Maybe some duplication can be saved by doing `impl<'a, T> Encodable for &'a [T] where T: AsRef<$type>`, which would cover both `&[T]` and `&[&T]`?

**Comment by sanket1729** [comment]
Correct me if I am wrong, this is just an internal refactor and does not add any new functionality.

File: src/consensus/encode.rs (hunk 1 above)
Comment (jrawsthorne) [reply src/consensus/encode.rs]: Fixed

File: src/consensus/encode.rs (hunk 2 above)
Comment (jrawsthorne) [reply src/consensus/encode.rs]: Fixed

File: src/consensus/encode.rs (hunk 3 above)
Comment (jrawsthorne) [reply src/consensus/encode.rs:549]: I would usually but I wasn't sure if we wanted to stick with only vec being encodable or expand it to include slices.

Comment (dpc) [reply src/consensus/encode.rs:549]: Taking `&[T]` is strictly better AFAIK, and taking `&Vec<T>` is like taking `&String` are 99.9% non-idiomatic. I thought there were `cargo clippy` error for it, but I might be wrong.

**Comment by sgeisler** [comment]
@sanket1729  I think `impl<'a> Encodable for Vec<&'a $type>` is new API, since outside users of the encodeing API might have needed to clone `Vec`s of references before to encode them.

Maybe we'd even want to do:

```rust
impl<'a, T, S> Encodable for S where T: AsRef<$type>, S: AsRef<[T]>
```

:sweat_smile: to allow encoding of `Vec<T>`, `Vec<&T>`, `&[T]` and `&[&T]`. But please write a test if you do so, that's a kinda wild impl. I think we want an owned encodable container type in case there is ever a need to write a trait bound like `Fn() -> E where E: Encodable`, where borrowed structures wouldn't work.

**Comment by sanket1729** [comment]
My bad. Should have read the description of the PR.

**Comment by Kixunil** [comment]
Note that blanket impls using `AsRef` are breaking change because someone might have implemented `Encodable for &[MyType]` and `AsRef<$type> for MyType`. (`Encodable` is not sealed.)

**Comment by tcharding** [comment]
Is this PR live still [jrawsthorne](https://github.com/jrawsthorne)?

**Comment by tcharding** [comment]
No movement on this for a few months, closing. Please re-open @jrawsthorne if this is still being worked on or anyone else if you want to see this functionality get in.

**Pull request closed without merging**